appmap/test/test_django.py

```diff
@@ -0,0 +1,23 @@
+import pytest
+from appmap._implementation.recording import Recorder
+
+pytest.importorskip('django')
+
+
+def test_sql_capture():
+    # pylint: disable=unused-import
+    import django.conf
+    import django.db
+    import appmap.django  # noqa: F401
+
+    recorder = Recorder()
+    recorder.events().clear()
+    recorder.enabled = True
+
+    django.conf.settings.configure(DATABASES={'default': {'ENGINE': 'django.db.backends.sqlite3', 'NAME': ':memory:'}})
+
+    conn = django.db.connections['default']
+    conn.cursor().execute('SELECT 1').fetchall()
+
+    recorder.enabled = False
+    assert recorder.events()[0].sql_query['sql'] == 'SELECT 1'
```

Comment:
```suggestion
    with recorder:
        django.conf.settings.configure(DATABASES={'default': {'ENGINE': 'django.db.backends.sqlite3', 'NAME': ':memory:'}})

        conn = django.db.connections['default']
        conn.cursor().execute('SELECT 1').fetchall()
    assert recorder.events()[0].sql_query['sql'] == 'SELECT 1'
```

Reply:
Will this clear the events list? I'm having trouble following the code there.

Reply:
Sorry, my mistake, this should actually start with
```
    from appmap import Recording
    ....
    recording = Recording()
    with recording:
```
`Recording` is a context manager, not `Recorder`, and it's intended to be a public interface.

Your question points to another problem, though: since `Recorder` is a singleton, it probably needs to protect `self._events` better. It should disallow multiple calls to `start_recording` with intervening calls to `stop_recording`, and it should clear out `_events` when `stop_recording` is called.

I don't think this test should need to worry about either of those cases, though. Are you seeing extra events in your appmap?

Reply:
The tests were broken in some orderings (see https://travis-ci.com/github/applandinc/appmap-python/builds/217319944 for instance). I'll see if I can fix that.

Reply:
Try putting your test in a class that inherits from `AppMapTestBase`. It has a `setup_method` member that properly reinitializes the modules.

Reply:
Ok, so another problem is that using `Recording()` needs an environment variable set. I can see you do that in the other tests, however this strikes me as rather brittle. I have to wonder: surely if you're doing `with Recording` as an external client (seeing as it's a public interface), you really do want to record, and be surprised if you need to set an environment variable for it to actually work?

Reply:
Yeah, that's probably true. But, let's leave it for now, and improve it going forward.

Reply:
And, using `Recording` also requires setting up a config file etc. It becomes more of an integration test that way. `Recording` is already tested with a separate test. Would it be okay to just leave this as it is?

Reply:
Yes, this is fine for now.